test/config/normalizers.test.js

```diff
@@ -0,0 +1,388 @@
+  const defaults = { badWithDefault: false }
+  const opts = {
+    boolTrue: true,
+    boolFalse: false,
+    strTrue: 'true',
+    strFalse: 'false',
+    badWithDefault: 'not-a-bool',
+    badWithoutDefault: 'not-a-bool',
+    nonBoolProperty: 25
+  }
+
+  normalizeBools(opts, fields, defaults, logger)
+
+  t.deepEqual(opts, {
+    boolTrue: true,
+    boolFalse: false,
+    strTrue: true,
+    strFalse: false,
+    badWithDefault: undefined,
+    badWithoutDefault: undefined,
+    nonBoolProperty: 25
+  })
+
+  const warnings = logger.calls
+  t.ok(warnings.length === 2, 'we got warnings for bad boolean options')
+  t.deepEqual(warnings[0].interpolation, ['not-a-bool', 'badWithDefault'])
+  t.deepEqual(warnings[1].interpolation, ['not-a-bool', 'badWithoutDefault'])
+  t.end()
+})
+
+test('#normalizeBytes()', function (t) {
+  const logger = new MockLogger()
+  const fields = ['bytes', 'kiloBytes', 'megaBytes', 'gigaBytes', 'numberBytes', 'badWithDefault', 'badWithoutDefault']
+  const defaults = { badWithDefault: '25kb' }
+  const opts = {
+    bytes: '1000b',
+    kiloBytes: '100kb',
+    megaBytes: '10mb',
+    gigaBytes: '1gb',
+    numberBytes: 12345678,
+    badWithDefault: 'not-bytes',
+    badWithoutDefault: 'not-bytes',
+    anotherProperty: 25
+  }
+
+  normalizeBytes(opts, fields, defaults, logger)
+
+  t.deepEqual(opts, {
+    bytes: 1000,
+    kiloBytes: 102400,
+    megaBytes: 10485760,
+    gigaBytes: 1073741824,
+    numberBytes: 12345678,
+    badWithDefault: NaN,
+    badWithoutDefault: NaN,
+    anotherProperty: 25
+  })
+
+  t.ok(logger.calls.length === 0, 'we got no warnings for bad byte options')
+  t.end()
+})
+
+test('#normalizeDurationOptions()', function (t) {
+  const logger = new MockLogger()
+  const fields = [
+    { name: 'withoutUnit', defaultUnit: 'ms', allowedUnits: ['us', 'ms', 's', 'm'], allowNegative: true },
+    { name: 'withUnit', defaultUnit: 'ms', allowedUnits: ['us', 'ms', 's', 'm'], allowNegative: true },
+    { name: 'notAllowedUnit', defaultUnit: 's', allowedUnits: ['s', 'm'], allowNegative: true },
+    { name: 'notAllowedNegative', defaultUnit: 's', allowedUnits: ['s', 'm'], allowNegative: false },
+    { name: 'badWithDefault', defaultUnit: 's', allowedUnits: ['s', 'm'], allowNegative: false },
+    { name: 'badWithoutDefault', defaultUnit: 's', allowedUnits: ['s', 'm'], allowNegative: false }
+
+  ]
+  const defaults = { badWithDefault: '25s' }
+  const opts = {
+    withoutUnit: '200',
+    withUnit: '2m',
+    notAllowedUnit: '20us',
+    notAllowedNegative: '-1s',
+    badWithDefault: 'not-duration',
+    badWithoutDefault: 'not-duration',
+    anotherProperty: 25
+  }
+
+  normalizeDurationOptions(opts, fields, defaults, logger)
+
+  t.deepEqual(opts, {
+    // TODO: this normalized deletes keys whose values are not valid (because format or contraint)
+    // other normalizers keep the property as is. Should we remove as well?
```

Comment:
review ☝️

Reply:
Per https://github.com/elastic/apm/blob/main/specs/agents/configuration.md#invalid-configuration-values we should be "ignoring" the value. Because of how we currently merge the different config sources with `Object.assign(...)` that means we should be deleting invalid values -- and logging a warning.

I think this PR should try *not to change behaviour* and fix these issues in a separate change.